### PR TITLE
feat(poetry): add POETRY_PYPI_TOKEN possibility as poetry credential

### PR DESCRIPTION
**Pull request by jhaumont**
## Changes

Adding a new env var to use a token as poetry credential for a private repo

## Context

Today, with poetry, we can use username/password possibility to set a credential for a private repository on Renovate (cf [doc](https://docs.renovatebot.com/getting-started/private-packages/#poetry)). Poetry also allow you to use a token, like explain in the [documentation](https://python-poetry.org/docs/repositories/#configuring-credentials:~:text=Alternatively%2C%20you%20can%20use%20environment%20variables%20to%20provide%20the%20credentials%3A). It could be interesting to support both solutions and not only username/password one.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Are you sure about this name and not `POETRY_PYPI_X_TOKEN`?

**Comment by jhaumont** [comment]
> Are you sure about this name and not `POETRY_PYPI_X_TOKEN`?

From what I understand on the documentation of Poetry, yes.
You have `poetry config http-basic.pypi <username> <password>` for basic auth and `poetry config pypi-token.pypi <my-token>` for a token. So for me, the kind is `PYPI_TOKEN` for a token and `HTTP_BASIC` for a basic auth. Not the more easier to understand ^^
I wait the validation of a python developper in my company regarding this (why I keep in draft mode for now)

**Comment by rarkins** [comment]
Unfortunately their documentation is a bit ambiguous due to the double use of PYPI:

image

Do you have the ability to _test_ this locally? Otherwise we're relying on unit tests and that gives us no way to know if it's right or not

File: docs/usage/getting-started/private-packages.md [hunk 1]
```diff
@@ -457,7 +457,7 @@ name = "pypi"
 
 ### poetry
 
-For every Poetry source, a `hostRules` search is done and then any found credentials are added to env like `POETRY_HTTP_BASIC_X_USERNAME` and `POETRY_HTTP_BASIC_X_PASSWORD`, where `X` represents the normalized name of the source in `pyproject.toml`.
+For every Poetry source, a `hostRules` search is done and then any found credentials are added to env like `POETRY_HTTP_BASIC_X_USERNAME` and `POETRY_HTTP_BASIC_X_PASSWORD`, or `POETRY_PYPI_TOKEN_X`, where `X` represents the normalized name of the source in `pyproject.toml`.
```
Comment (viceice) [inline docs/usage/getting-started/private-packages.md:460]:
```suggestion
For every Poetry source, a `hostRules` search is done and then any found credentials are added to env like `POETRY_HTTP_BASIC_X_USERNAME` and `POETRY_HTTP_BASIC_X_PASSWORD`, or `POETRY_HTTP_TOKEN_X`, where `X` represents the normalized name of the source in `pyproject.toml`.
```
please align naming to existing environment variables

Comment (jhaumont) [reply docs/usage/getting-started/private-packages.md:460]: Hello! 
This is not a choice, this naming is required by poetry it self: POETRY_PYPI_X_TOKEN
cf https://python-poetry.org/docs/repositories/#configuring-credentials:~:text=Alternatively%2C%20you%20can%20use%20environment%20variables%20to%20provide%20the%20credentials%3A

**Comment by jhaumont** [comment]
> Unfortunately their documentation is a bit ambiguous due to the double use of PYPI:
> 
> image
> 
> Do you have the ability to _test_ this locally? Otherwise we're relying on unit tests and that gives us no way to know if it's right or not

Yes, we have the same feeling than you. We try to test it with artifact registry on GCP but in fact... token can work with this registry.
You can only use a keyring or a service account (cf [official doc](https://cloud.google.com/artifact-registry/docs/python/authentication)).
We try to find another way to do the test.

Also, on our side we will test the possibility to set a service account key to use the already existing method on renovate with basic auth for poetry.

Do you think it could be a possibility to add the [keyring auth method](https://cloud.google.com/artifact-registry/docs/python/authentication#keyring-setup) for poetry on renovate or it's too complicated ? (for me, it's too complicated :p )

**Comment by jhaumont** [comment]
@rarkins So finally, we have found a solution using the basic auth mecanism already implement in renovate (cf this [doc](https://cloud.google.com/artifact-registry/docs/python/authentication#sa-key)). So I propose to close this PR.

**Pull request closed without merging**